File: src/developer/forensics/feedback_data/screenshot.cc
// Copyright 2019 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "src/developer/forensics/feedback_data/screenshot.h"

#include <fuchsia/ui/composition/cpp/fidl.h>

#include "src/developer/forensics/utils/errors.h"
#include "src/developer/forensics/utils/fidl_oneshot.h"

namespace forensics::feedback_data {

::fpromise::promise<ScreenshotData, Error> TakeScreenshot(
    async_dispatcher_t* dispatcher, std::shared_ptr<sys::ServiceDirectory> services,
    zx::duration timeout) {
  fuchsia::ui::composition::ScreenshotTakeRequest args;
  args.set_format(fuchsia::ui::composition::ScreenshotFormat::BGRA_RAW);
  return OneShotCall<fuchsia::ui::composition::Screenshot,
                     &fuchsia::ui::composition::Screenshot::Take,
                     fuchsia::ui::composition::ScreenshotTakeRequest>(dispatcher, services, timeout,
                                                                      std::move(args))
      .and_then([](fuchsia::ui::composition::ScreenshotTakeResponse& result)
                    -> ::fpromise::result<ScreenshotData, Error> {
        ScreenshotData data;
        data.info.transform = fuchsia::images::Transform::NORMAL;
        data.info.width = result.size().width;
        data.info.height = result.size().height;
        data.info.stride = data.info.width * 4;
        data.info.pixel_format = fuchsia::images::PixelFormat::BGRA_8;
        data.info.color_space = fuchsia::images::ColorSpace::SRGB;
        data.info.tiling = fuchsia::images::Tiling::LINEAR;
        data.info.alpha_format = fuchsia::images::AlphaFormat::OPAQUE;
        data.data =
            fsl::SizedVmo(std::move(*result.mutable_vmo()), data.info.height * data.info.stride);
        return ::fpromise::ok(std::move(data));
      });
}

}  // namespace forensics::feedback_data

File: src/testing/end_to_end/honeydew/honeydew/affordances/sl4f/ui/screenshot.py
# Copyright 2023 The Fuchsia Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.
"""Screenshot affordance implementation using SL4F."""

import base64
from typing import Any

from honeydew import errors
from honeydew.interfaces.affordances.ui import custom_types, screenshot
from honeydew.transports import sl4f as sl4f_transport

_SL4F_METHODS: dict[str, str] = {
    "Take": "scenic_facade.TakeScreenshot",
}


class Screenshot(screenshot.Screenshot):
    """Screenshot affordance implementation using SL4F.

    Args:
        sl4f: SL4F transport.
    """

    def __init__(self, sl4f: sl4f_transport.SL4F) -> None:
        self._sl4f: sl4f_transport.SL4F = sl4f

    def take(self) -> screenshot.ScreenshotImage:
        """Take a screenshot.

        Return:
            ScreenshotImage: the screenshot image.

        Raise:
            Sl4fError: sl4f responses in unexpected format.
        """
        resp: dict[str, Any] = self._sl4f.run(method=_SL4F_METHODS["Take"])

        try:
            width: int = resp["result"]["info"]["width"]
            height: int = resp["result"]["info"]["height"]
            img_bytes = resp["result"]["data"].encode("utf-8")
        except KeyError as err:
            raise errors.Sl4fError(err)

        data: bytes = base64.decodebytes(img_bytes)
        return screenshot.ScreenshotImage(
            size=custom_types.Size(width=width, height=height), data=data
        )
